packages/graphback-datasync/src/providers/DeltaProvider.ts

```diff
@@ -0,0 +1,203 @@
+import { applyIndexes } from "@graphback/runtime-mongo";
+import { GraphQLObjectType } from 'graphql';
+import { Db, ObjectId, FilterQuery } from "mongodb"
+import { getTableName, GraphbackOperationType } from '@graphback/core';
+
+function getDeltaTableName(baseType: GraphQLObjectType) {
+  return `Graphback-${getTableName(baseType)}-Delta`
+}
+
+type DeltaOps = GraphbackOperationType.CREATE | GraphbackOperationType.UPDATE | GraphbackOperationType.DELETE;
+
+export interface ConflictingFieldsReport {
+  _id: ObjectId,
+  conflicts: {
+    [fieldName: string]: any
+  },
+  version: number
+}
+
+interface DeltaDocument {
+  docId: ObjectId,
+  op: DeltaOps,
+  timestamp: Date,
+  version: number,
+  sets: {
+    [fieldName: string]: any
+  }
+}
+
+
+/**
+ * Provides an interface for interacting with deltas and stuff
+ */
+export class DeltaProvider {
+  protected collectionName: string;
+  protected db: Db;
+
+  public constructor(baseType: GraphQLObjectType, db: Db) {
+    this.db = db;
+
+    this.collectionName = getDeltaTableName(baseType);
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    applyIndexes([
+      {
+        key: {
+          "docId": 1,
+          "version": 1
+        }
+      }
+    ], this.db.collection(this.collectionName));
+  }
+
+
+  public async insertDiff(_id: string|ObjectId, ts: string|number, documentVersion: number, updateData: any, op: DeltaOps) {
+    const docId = getObjectIdFromString(_id);
+    const timestamp = getDateFromTimestamp(ts);
+    const sets = getDiffEntry(updateData);
+
+    const deltaDocument: DeltaDocument = {
+      op,
+      docId,
+      timestamp,
+      version: documentVersion,
+      sets
+    }
+
+    const res = await this.db.collection(this.collectionName).insertOne(deltaDocument);
+
+    return res.insertedId;
+  }
+
+  public async sync(lastSync: number, filter?: FilterQuery<any>) {
```

Comment:
Any reason for using delta table for reading/sync
This is:
- Very slow (Probably)
- Complex
- Hard to maintain

We have the same information in the main table (with soft deletes). We do not need to remove soft deletes IMHO.
We keep delta table for conflict resolution (with some TTL)

Using it for delta requires complex merges that can span many layers and even return incomplete data (if you will miss entire objects or you start removing objects by TTL.)

Generally with some aggressive TTL we cannot avoid having situation that client will need to handle conflict (possibly by reapplying change etc.

Reply:
This is going away

Would like to chat about what would happen if delta document deleted by TTL, kinda lost there on what to do when that happens

Reply:
No need to be lost. TTL is simply to minimize amount of space. If your record hit TTL you cannot perform conflict resolution on sever and return error to user. 

No matter if you are getting partial results (some of the elements in common history) - if you do not have common base we cannot realize server side conflict.

In practical means - if I made edit when offline 5 min before going on 2 weeks holiday there is no sense to automatically resolve that change on server - History will be most likely gone and objects might have moved so much that using auto strategy on server will cause havoc in this cases. Returning error to user will let him to apply their changes again.